### PR TITLE
[SYCL][FPGA] Ignore duplicate merge attributes applied to declarations with different parameters

### PR DESCRIPTION
**Pull request by smanna12**
Fixes https://github.com/intel/llvm/issues/7818

### CODE REVIEW

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 1]
```diff
@@ -298,10 +298,19 @@ void diagnostics()
...
   // Checking of different argument values.
-  //expected-warning@+2{{attribute 'merge' is already applied}}
-  [[intel::merge("mrg4", "depth")]]
+  //expected-warning@+2{{attribute 'merge' is already applied with different arguments}}
+  [[intel::merge("mrg4", "depth")]] // expected-note {{previous attribute is here}}
   [[intel::merge("mrg5", "width")]] unsigned int mrg_mrg[4];
 
+  // Checking of duplicate argument values.
+  // No diagnostic is emitted because the arguments match.
+  //CHECK: VarDecl{{.*}}mrg_mrg6
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/intel-fpga-local.cpp:307]: Isn't this a diagnostics test? I don't think you need to check AST here again

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 2]
```diff
@@ -298,10 +298,19 @@ void diagnostics()
   [[intel::merge("mrg4", "depths")]] unsigned int mrg_invalid_arg[4];
 
   // Checking of different argument values.
-  //expected-warning@+2{{attribute 'merge' is already applied}}
-  [[intel::merge("mrg4", "depth")]]
+  //expected-warning@+2{{attribute 'merge' is already applied with different arguments}}
+  [[intel::merge("mrg4", "depth")]] // expected-note {{previous attribute is here}}
   [[intel::merge("mrg5", "width")]] unsigned int mrg_mrg[4];
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/intel-fpga-local.cpp:303]: Can you add a test where the first argument matches but not the second?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 3]
```diff
@@ -7588,8 +7588,6 @@ static void handleSYCLIntelMaxReplicatesAttr(Sema &S, Decl *D,
 /// second is a direction.  The direction must be "depth" or "width".
 /// This is incompatible with the register attribute.
 static void handleSYCLIntelMergeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
-  checkForDuplicateAttribute<SYCLIntelMergeAttr>(S, D, AL);
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaDeclAttr.cpp:7591]: Can we get rid of this? Or is it being called elsewhere? Is this community code?

File: clang/test/SemaSYCL/intel-fpga-local.cpp (hunk 2 above)
Comment (smanna12) [reply clang/test/SemaSYCL/intel-fpga-local.cpp:303]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 3 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:7591]: This is not a community code. We added this part of FPGA attribute support. IT has been used by other FPGA attributes. 
```
checkForDuplicateAttribute<SYCLIntelSinglePumpAttr>
checkForDuplicateAttribute<SYCLIntelDoublePumpAttr>
checkForDuplicateAttribute<SYCLIntelMemoryAttr>
checkForDuplicateAttribute<SYCLIntelRegisterAttr>
checkForDuplicateAttribute<SYCLIntelSimpleDualPortAttr>
checkForDuplicateAttribute<SYCLIntelBankBitsAttr>

```
I will remove it once we fix duplicate attributes bugs with them.

File: clang/test/SemaSYCL/intel-fpga-local.cpp (hunk 1 above)
Comment (smanna12) [reply clang/test/SemaSYCL/intel-fpga-local.cpp]: My mistake. Thanks @elizabethandrews for reviews and catching it. I have removed it.

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 4]
```diff
@@ -298,10 +298,18 @@ void diagnostics()
   [[intel::merge("mrg4", "depths")]] unsigned int mrg_invalid_arg[4];
 
   // Checking of different argument values.
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/intel-fpga-local.cpp:300]: Sorry I didn't mention this earlier but can you also add a test where arguments match. I know we have an AST test for this but it makes sense to endure diagnostics work as expected in this case too.

Comment (smanna12) [reply clang/test/SemaSYCL/intel-fpga-local.cpp:300]: Thanks @elizabethandrews for reviews and feedbacks. 

>>can you also add a test where arguments match.

Yes, i have added the test.

**Review by elizabethandrews — APPROVED** [review]

**Comment by smanna12** [comment]
@intel/llvm-gatekeepers, This PR is ready to be merged? Thanks!

**Pull request merged**